core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -0,0 +1,288 @@
+    aclChangeListener.startup()
+
+    zkClient.subscribeStateChanges(ZkStateChangeListener)
+
+    loadCache
+  }
+
+  override def authorize(session: Session, operation: Operation, resource: Resource): Boolean = {
+    val principal: KafkaPrincipal = session.principal
+    val host = session.host
+
+    if (superUsers.contains(principal)) {
+      authorizerLogger.debug(s"principal = $principal is a super user, allowing operation without checking acls.")
+      logAuditMessage(principal, authorized = true, operation, resource, host)
+      return true
+    }
+
+    val acls: Set[Acl] = getAcls(resource)
+    if (acls.isEmpty) {
+      authorizerLogger.debug(s"No acl found for resource $resource, authorized = $shouldAllowEveryoneIfNoAclIsFound")
+      logAuditMessage(principal, authorized = shouldAllowEveryoneIfNoAclIsFound, operation, resource, host)
+      return shouldAllowEveryoneIfNoAclIsFound
+    }
+
+    //first check if there is any Deny acl match that would disallow this operation.
+    if (aclMatch(session, operation, resource, principal, host, Deny, acls)) {
+      logAuditMessage(principal, authorized = false, operation, resource, host)
+      return false
+    }
+
+    /**
+     * if principal is allowed to read or write we allow describe by default, the reverse does not apply to Deny.
+     */
+    val ops: Set[Operation] = if (Describe == operation)
+      Set[Operation](operation, Read, Write)
+    else
+      Set[Operation](operation)
+
+    //now check if there is any allow acl that will allow this operation.
+    if (ops.exists(operation => aclMatch(session, operation, resource, principal, host, Allow, acls))) {
+      logAuditMessage(principal, authorized = true, operation, resource, host)
+      return true
+    }
+
+    //We have some acls defined and they do not specify any allow ACL for the current session, reject request.
+    authorizerLogger.debug(s"principal = $principal from host = $host is not allowed to perform operation = $operation  " +
+      s"on resource = $resource as no explicit acls were defined to allow the operation")
+    logAuditMessage(principal, authorized = false, operation, resource, host)
+    false
+  }
+
+  private def aclMatch(session: Session, operations: Operation, resource: Resource, principal: KafkaPrincipal, host: String, permissionType: PermissionType, acls: Set[Acl]): Boolean = {
+    val acl = acls.find ( acl =>
+      acl.permissionType == permissionType
+      && (acl.principal == principal || acl.principal == Acl.WildCardPrincipal)
+      && (operations == acl.operation || acl.operation == All)
+      && (acl.host == host || acl.host == Acl.WildCardHost)
+    ).map{ acl: Acl =>
+      debug(s"operation = $operations on resource = $resource from host = $host is $permissionType based on acl = $acl")
+      return true
+    }
```

Comment:
I agree that this version looks worse. It's not what I suggested. :) A `getOrElse` after the map is what I would do. Do you still prefer the previous version in that case?